### PR TITLE
Compiler work

### PR DESCRIPTION
**Pull request by AWoloszyn**

### CODE REVIEW

File: gapil/compiler/map.go [hunk 1]
```diff
@@ -335,6 +335,7 @@ func (c *compiler) buildMapType(t *semantic.Map) {
 
 	c.build(mi.Lookup, func(s *scope) {
 		m := s.Parameter(0).SetName("map")
+		s.arena = m.Index(0, mapArena).Load().SetName("arena")
```
Comment (ben-clayton) [inline gapil/compiler/map.go]: Nit: please move one line down. Not only is this consistent with the other functions, it'll save me a merge conflict when I rebase the protos branch.

File: core/codegen/builder.go [hunk 2]
```diff
@@ -281,3 +281,8 @@ func (b *Builder) block(block, next llvm.BasicBlock, f func()) {
 		b.llvm.CreateBr(next)
 	}
 }
+
+// Dumps the current module to stderr.
+func (b *Builder) Dump() {
```
Comment (ben-clayton) [inline core/codegen/builder.go]: Belongs on `Module`. Note there is also `Module.String()`

File: gapil/compiler/statements.go [hunk 3]
```diff
@@ -346,11 +346,15 @@ func (c *compiler) read(s *scope, n *semantic.Read) {
...
+	if n.Value != nil {
+		val := c.expression(s, n.Value)
+		c.reference(s, val, n.Value.ExpressionType())
 
-	retTy := c.returnType(c.currentFunc) // <error, value>
-	c.doReturn(s, s.Zero(retTy).Insert(retValue, val))
+		retTy := c.returnType(c.currentFunc) // <error, value>
+		c.doReturn(s, s.Zero(retTy).Insert(retValue, val))
+	} else {
+		c.doReturn(s, nil)
```
Comment (ben-clayton) [inline gapil/compiler/statements.go]: Relying here on `<0, 0>` being an acceptable return value, when these aren't legal values for all types.

The `string`, `class` and `map` type's `initialValue`s  are not the same as an `codegen.Builder.Zero()` values. Returning a zero value for these will likely cause NPEs for these types as the compiler expects values to be non-null.

We probably want:

```go
var val *codegen.Value
if n.Value != nil {
	val = c.expression(s, n.Value)
} else {
	val = c.initialValue(s, n.Value.ExpressionType())
}
c.reference(s, val, n.Value.ExpressionType())
retTy := c.returnType(c.currentFunc) // <error, value>
ret := s.Zero(retTy).Insert(retValue, val)
c.doReturn(s, ret)
```

Note: I also made the same mistake in the proto branch.

File: gapil/compiler/map.go (hunk 1 above)
Comment (AWoloszyn) [reply gapil/compiler/map.go]: Done

File: core/codegen/builder.go (hunk 2 above)
Comment (AWoloszyn) [reply core/codegen/builder.go]: I can remove, this was useful for testing, but probably not something worth keeping in the API anyway.

File: gapil/compiler/statements.go (hunk 3 above)
Comment (AWoloszyn) [reply gapil/compiler/statements.go]: Done

Comment (AWoloszyn) [reply gapil/compiler/statements.go]: `val = c.initialValue(s, n.Value.ExpressionType())` Won't work, but I have an update that should fix it.

**Review by ben-clayton — APPROVED** [review]
Awesome, thanks!

**Comment by AWoloszyn** [comment]
Thanks! (I cannot however actually merge)

**Pull request merged**